sklearn/model_selection/tests/test_validation.py

```diff
@@ -1259,61 +1261,220 @@ def test_cross_val_predict_sparse_prediction():
+    X, y = make_classification(n_classes=2, random_state=0)
+    for method in ['decision_function', 'predict_proba', 'predict_log_proba']:
+        est = LogisticRegression()
+        check_cross_val_predict_with_method_binary(est, X, y, method)
 
 
-def test_cross_val_predict_with_method():
-    check_cross_val_predict_with_method(LogisticRegression())
+def test_cross_val_predict_with_method_multiclass():
+    iris = load_iris()
+    X, y = iris.data, iris.target
+    X, y = shuffle(X, y, random_state=0)
+    for method in ['decision_function', 'predict_proba', 'predict_log_proba']:
+        est = LogisticRegression()
+        check_cross_val_predict_with_method_multiclass(est, X, y, method)
 
 
 def test_cross_val_predict_method_checking():
     # Regression test for issue #9639. Tests that cross_val_predict does not
     # check estimator methods (e.g. predict_proba) before fitting
-    est = SGDClassifier(loss='log', random_state=2)
-    check_cross_val_predict_with_method(est)
+    iris = load_iris()
+    X, y = iris.data, iris.target
+    X, y = shuffle(X, y, random_state=0)
+    for method in ['decision_function', 'predict_proba', 'predict_log_proba']:
+        est = SGDClassifier(loss='log', random_state=2)
+        check_cross_val_predict_with_method_multiclass(est, X, y, method)
 
 
 def test_gridsearchcv_cross_val_predict_with_method():
+    iris = load_iris()
+    X, y = iris.data, iris.target
+    X, y = shuffle(X, y, random_state=0)
     est = GridSearchCV(LogisticRegression(random_state=42),
                        {'C': [0.1, 1]},
                        cv=2)
-    check_cross_val_predict_with_method(est)
+    for method in ['decision_function', 'predict_proba', 'predict_log_proba']:
+        check_cross_val_predict_with_method_multiclass(est, X, y, method)
+
+
+def test_cross_val_predict_with_method_multilabel_ovr():
+    # OVR does multilabel predictions, but only arrays of
+    # binary indicator columns. The output of predict_proba
+    # is a 2D array with shape (n_samples, n_classes).
+    n_samp = 100
+    n_classes = 4
+    X, y = make_multilabel_classification(n_samples=n_samp, n_labels=3,
+                                          n_classes=n_classes, n_features=5,
+                                          random_state=42)
+    est = OneVsRestClassifier(LogisticRegression(random_state=0))
+    for method in ['predict_proba', 'decision_function']:
+        check_cross_val_predict_with_method_binary(est, X, y, method=method)
+
+
+class RFWithDecisionFunction(RandomForestClassifier):
+    # None of the current multioutput-multiclass estimators have
+    # decision function methods. Create a mock decision function
+    # to test the cross_val_predict function's handling of this case.
+    def decision_function(self, X):
+        probs = self.predict_proba(X)
+        if isinstance(probs, list):
+            probs = [p[:, -1] if p.shape[1] == 2 else p for p in probs]
+        elif probs.shape[1] == 2:
```

Comment:
I was a bit nervous about removing this branch. This class only exists to help me test `decision_function` for list outputs, but I thought it was better to have the mock `decision_function` not be broken for non-list outputs. Would you still prefer either that this branch turn into a `ValueError`, or that I assume it will always run with list outputs?

Reply:
Just put assert isinstance before and get rid of this case

Reply:
Okay, done.